### PR TITLE
fix(KVS): - Add Support for More Resolution Widths

### PR DESCRIPTION
**Pull request by stefankiesz**
## Issue
When streaming to Kinesis Video Streams using this SDK, the resolution of the frame Image read from the Android camera may not match up with the resolution configured by the user's application, causing completely distorted frames to be ingested:

image

This is due to a standard memory layout optimization when processing images. Typically, an image's width must be a multiple of a value such as, for example, 64 in order to line up with the optimal line/row stride of the image. If the image width is not a multiple of 64, padding is added to get the width up to the nearest multiple of 64. However, the encoder would still be configured for the user-defined resolution's width, thus causing such image stride distortions.

From my testing with the [AmazonKinesisVideoDemoApp](https://github.com/awslabs/aws-sdk-android-samples/tree/main/AmazonKinesisVideoDemoApp) on a Samsung Galaxy S22 Ultra, the selected resolution must indeed be a multiple of 64 in order for frames to come into KVS undistorted. Here are some results:

176 x 144 - broken
320 x 240 - good
352 x 288 - broken
640 x 360 - good
640 x 480 - good
720 x 480 - broken
960 x 720 - good
1088 x 1088 - good
1280 x 720 - good
1440 x 1080 - broken
1920 x 824 - good
1920 x 1080 - good
1440 x 1080 - broken

Now with some more details added, the table below shows the frame width mismatch between the output buffers of the Images coming in from the camera and the encoder's input buffers:

image

## Solution
The solution here is to manually check for whether the Image rowStride is not equivalent to the Image width. If not equivalent, then ignore the padding when copying the ByteBuffers from the outputted Image to the encoder's input ByteBuffers. So, in the case where there is image padding to ignore, for each frame, we have to make _h_ copy operations instead of just 1 where _h_ is the height of the image (_h_ is the row count).

## Testing
These changes were tested by building this SDK locally and running with the [AmazonKinesisVideoDemoApp](https://github.com/awslabs/aws-sdk-android-samples/tree/main/AmazonKinesisVideoDemoApp). The video footage was ingested without distortion for all tested resolutions. There was a slight drop in frame rate for the DemoApp's video preview, likely due to the additional copy operations.

<br>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/mobileconnectors/kinesisvideo/encoding/EncoderFrameSubmitter.java
Comment (sirknightj) [inline aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/mobileconnectors/kinesisvideo/encoding/EncoderFrameSubmitter.java:144]:
```suggestion
        final boolean isRowPaddingPresent = imageWidth != imageRowStride;
```

File: aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/mobileconnectors/kinesisvideo/encoding/EncoderFrameSubmitter.java
Comment (mattcreaser) [inline aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/mobileconnectors/kinesisvideo/encoding/EncoderFrameSubmitter.java:145]:
```suggestion
        final boolean isSemiPlanar = imageChromaPixelStride != 1;
```

File: aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/mobileconnectors/kinesisvideo/encoding/EncoderFrameSubmitter.java
Comment (mattcreaser) [inline aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/mobileconnectors/kinesisvideo/encoding/EncoderFrameSubmitter.java:180]:
```suggestion
                                 final int imageWidth, 
                                 final int sourceBufferRowStride, 
                                 final int pixelStride) {
```

**Review by mattcreaser — COMMENTED** [review]
Thanks for the well-commented code, this is easy to follow. Just a couple of small syntax cleanups needed.

**Comment by stefankiesz** [comment]
@sirknightj @mattcreaser, thanks for taking a look folks! I just merged in all your suggested changes.

File: aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/mobileconnectors/kinesisvideo/encoding/EncoderFrameSubmitter.java
Comment (sirknightj) [inline aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/mobileconnectors/kinesisvideo/encoding/EncoderFrameSubmitter.java:156]:
```suggestion
            if (isRowPaddingPresent && isSemiPlanar) {
```

Comment (stefankiesz) [reply aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/mobileconnectors/kinesisvideo/encoding/EncoderFrameSubmitter.java]: Nice catch!

**Review by sirknightj — APPROVED** [review]

**Review by mattcreaser — APPROVED** [review]

**Comment by stefankiesz** [comment]
Looks like we need 1 more maintainer approval for merging whenever anyone gets the chance, thanks.

**Review by vincetran — APPROVED** [review]

**Pull request merged**
